### ISSUE TITLE
Add an `aka` for the identity of a core

### ISSUE THREAD

**Issue by chc4** [urbit/arvo#361]
You've surely seen those arcane typecasts of `^-  {(list move) _+>.$}` in state machines and app arms. It is, frankly, scary and hard to explain since it's basically the equivalent of dereferencing a pointer literal in C, and pretty much the only spot where Nock shines through. It's basically the only reason for the explanation of lark syntax.

`=*` allows you to define a variable as an alias instead of by-value copy. They also update through context updates (`(|=(@ =*(b +< =+(c=1 b))) 4)` compared to `(|=(@ =+(c=1 +<)) 4)`), unlike the current `+>.$` invocation which you have to potentially manually update if you add a pin.

Is there a reason that `|_` at least doesn't add a `=*  this  +>.$` to its expansion? I don't actually know how %core is expanded, so it might require something like adding the bull to the top of each arm in the core the door expands into, since the sub-twig of a core is actually a `(map term twig)` instead.

**Comment by joemfb** [comment]
Haven't tested anything yet, but I think something like this would work:

```
[%aka p=%sample q=[%make p=~[[%.y p=2]] q=~] r=[%aka p=%context q=[%make p=~[[%.y p=3]] q=~] r=[%core ...]]]
```

**Comment by chc4** [comment]
Maybe? `+>.$` is axis 7 inside the core, while that's defining the axis alias outside of it. Depends on how the expansion of aka fixes up the axis so that `(|=(@ =*(b +< =+(c=1 b))) 4)` works correctly, because it might fix it so that it now points to what axis 2 and 3 are outside.

It's makes me feel vaguely dirty just looking at it, since it's like you're defining a reference to a variable before the variable is defined, but if it works that'd be great.

**Comment by joemfb** [comment]
This would be *inside* the sample `%pin`.

**Comment by joemfb** [comment]
In other words:

```
=<  ~(foo . 3)  =|  a/@  =*  sample  a  =*  context  +3  |%  ++  foo  sample  --
```

**Comment by joemfb** [comment]
```
> (ream '=+  *a/@  =*  arg  ,.+2  =*  this  ,.+3  |%  ++  foo  arg  --')
[ %pin
  p=[%bunt p=[%coat p=%a q=[%base p=[%atom p=~.]]]]
  q=[%aka p=%arg q=[%wing p=~[[%.n p=0 q=~] [%.y p=2]]] r=[%aka p=%this q=[%wing p=~[[%.n p=0 q=~] [%.y p=3]]] r=[%core p={[p=%foo q=[%ash p=[%wing p=~[%arg]]]]}]]]
]
```

**Comment by chc4** [comment]
Ah, you're right. That does make this a lot easier then, and so it's basically just bikeshedding.

Mentioned in `:talk` was that `..foo` is the same as `+>.$`, but works even with pins. Problems are that it's not actually clear what that means (and the expression _..foo doesn't actually have anything to do with the ++foo arm), along with breaking via spurious renames of unrelated arms. You also have to pick an arbitrary arm while you're coding, which is kinda kludgy. These is, admittedly, minor gripes, and magically defining a magic variable that names the door might not be any better?

/cc benevolent dictator @cgyarvin

**Comment by joemfb** [comment]
No, this is wrong, and you were right in `urbit-meta`. `+3` is the context of the door, not the door itself. Each arm would need to be wrapped in the equivalent of `=*(this . ...)` ...

**Comment by joemfb** [comment]
In `++boil:ap`:

```
{$door *}
  :+  %pin  [%bunt p.gen]
  :-  %core
  %-  ~(run by q.gen)
  |=(a/twig [%aka p=%this q=[%wing p=~[[| p=0 q=~] [& p=1]]] r=a))
```

**Comment by cgyarvin** [comment]
I don't like the magic variable either, that's very un-Hoon (well, I guess
we do have %abel and %cain, but they stay hidden).  Nor do I like ..foo.

What I often do is just, instead of |_,

=|  state
=*  sys  -
|%
...
--

This has exactly the effect you want, though I admit it's a little wordy.
But I am not sure this pattern is crying out for syntactic help.

**Comment by joemfb** [comment]
That gets you easy access to the full sample, but not easy access to the core itself. The magic face is certainly not Hoonish, but state-machine dores can be pretty hard to follow ...

**Comment by cgyarvin** [comment]
Yeah.  I agree that there's no particularly likable solution here.

If we are willing to defile %core with this change, which is kind of a pain
in the ass but not *that* hard, one could imagine adding a (unit twig)
which served as a general arm prelude, being automagically =>'ed onto every
arm in the core.  This is where you'd put any number of =* aliases, and
perhaps even other things.

**Comment by chc4** [comment]
I feel better if I think about it like the "magic" `$` arm name. But it is definitely very un-Hoon.

Being able to write `|_  a/@  ++  foo  |=  b/@  ^-  {@  _this}  [(add a b) this(a (add a b))]` would be pretty great, though, signaling that state machines are a Feature(tm). `..foo` does work nearly just as well, though.

Re: (unit twig) prelude, I'd actually love that. I wanted something similar for access controls for app pokes, for example, instead of having ?:  (team who own) at every one, and after playing around for a while wasn't able to get anything working.

**Comment by cgyarvin** [comment]
I would certainly love a PR if anyone feels capable of adding this feature.

**Comment by joemfb** [comment]
Trying to picture, would it be a new core syntax?

```
=<  abet:(~(foo . 10) 5)
|#  a/@
  =*  this  .
++  abet
  [~  a]
++  foo
  |=  b/@
  this(a (add a b))
--
```

**Comment by joemfb** [comment]
That would obviously require the ability to partially parse twigs

**Comment by cgyarvin** [comment]
It would be syntactically a peer of ++.

**Comment by hoclun-rigsep** [comment]
Do we want to go ahead with the `(unit twig)` idea? Should we begin arguing about the rune? I propose `+:`

I believe I can implement this.

**Comment by juped** [comment]
Yeah. Defile away, hoclun!

**Comment by ohAitch** [comment]
For the +> case in particular, how much is covered by just saying ++ self .
in all your cores? I've used that in a couple of places, thought it was
definitely harder to read than ..foo

**Comment by cgyarvin** [comment]
Sounds great.  (I would probably pick +* or +/, but it doesn't matter
enough to violate the He Who Does The Work principle).

Bear in mind that since you're changing the shape of ++span, you're going
to have to dabble with the jets as well.  Elliot already did this, he may
be able to advise...

**Comment by cgyarvin** [comment]
That's also not a perfect substitute, as you can't say self(foo bar).  You
have to say ..self(foo bar).  Which is not hugely onerous, but is rather
weird.  Also, you generate this unnecessary arm, which is inelegant.

**Comment by juped** [comment]
I'm just a simple country mathematician, but how does this change spans?

**Comment by ohAitch** [comment]
You can in fact say self(foo bar), it evaluates to . ie the whole core.

Yeah it's not better than the magically autoinserted arm, but if we're
putting in a line in every core anyway it seems not worse; but this goes
back to the tradeoff between compile-time-only magic vs having everything
you do be actual nouns w/ nock in.

**Comment by cgyarvin** [comment]
Yeah okay, you can say self(foo bar), but it still evaluates in a really
weird way.  I'm not saying you can't do this, but it's just gnarly as a
mainstream idiom.

**Comment by cgyarvin** [comment]
Because (I would expect) this changes ++coil to be {p/?($gold $iron $lead
$zinc) q/span r/(unit twig) s/{p/?($~ ^) q/(map term foot)}.

Unless you're thinking of applying the prelude right there in the parser?
Well, okay, I could see this.  It seems somehow tacky, though.

**Comment by ohAitch** [comment]
I'm not sure how I see +: =* self . ++ etc self(foo bar) being any less
weird? Mostly you're just putting in a new place where names can be bound.

**Comment by cgyarvin** [comment]
You're not adding a superfluous arm.  There is a huge division between compile time hacks and runtime hacks.

**Comment by ohAitch** [comment]
That I agree on, the "rewrite all the arms" solution is more weird and
harder to unravel, but probably uses slightly fewer nock reductions P:

**Comment by hoclun-rigsep** [comment]
I am getting cold feet about whether we *should* defile `%core`. Fwiw I disagree with the OP motivation of adding "this" as, effectively, a keyword. 

@chc4, can you elaborate on the use case you mentioned: https://github.com/urbit/arvo/issues/361#issuecomment-283857933

**Comment by chc4** [comment]
Right now, all ++poke-noun arms are by default public and the only way to have then do security checking is by adding an assertion at the top of each one testing that the poker is authorized to do that. Having an arm prelude would let you apply that to all the arms at once, and since cores are concatative you could have two cores internally: one which is public API, and one which is private API.

Automagically adding a `++this  .` arm would actually be more hoonish, now that I think of it. There is precedent with `++$` being put everywhere, and there's not a need to add a mechanism to support it (just put:by the arm in the door case). I'm pretty sure the current system doesn't allow you to only partially parse a twig, so adding a `+:` arm type would be hacky at best.

**Comment by joemfb** [comment]
Applying a partial twig to each arm could be cool, but you'd have to hack it or rewrite the parser. Anyway, it was my misreading of @cgyarvin's proposal -- he just said to => it on each arm.

Ex:

```
|_  a/@
+*  =*(this . .)
++  abet  [~ a]
++  flub
  |=  b/@
  this(a (add a b))
--
```

**Comment by cgyarvin** [comment]
Yeah.  I don't see a huge reason, thinking about it, to not apply this
prelude in the parser -- saving any surgery on ++span.

Frankly, this is the kind of wart languages start growing as they mature.
It shouldn't need to be there but it does need to be there.  So someone
adds it.

**Comment by ohAitch** [comment]
Okay "+* =*(this . .)" is just straightforwardly more verbose than "++
 this  .", which doesn't even require any compiler changes. Is the extra
nock 9 that big of a deal, or am I missing some higher principle?

**Comment by ohAitch** [comment]
Like, one way to split the difference would be to have +* be basically
__attr(inline)__ ++, usage

|%
+* this .
++  x  this(foo bar)
--

Which can then be implemented via =* at the top of every arm, or even in
the core type: "this is the set of named nock formulas available in this
noun, as distinct from arms which are also physically present in this noun"

But this seems like a lot of increased abstraction complexity, for
something that has not yet been determined to be any sort of performance
bottleneck.

**Comment by joemfb** [comment]
I've pulled together 3 variants of this proposal. (For the last two, I added `$dork` rather than figure out how to reset/solid a twig page change ...)

### magic face

https://github.com/urbit/arvo/compare/maint-20160818...joemfb:door-this

```
|=  arg/@
=<  abet:(~(flub . 10) arg)
|_  a/@
++  abet  [~ a]
++  flub
  |=  b/@
  this(a (add a b))
--
```

```
> +test 10
[~ 20]
```

### custom subject

https://github.com/urbit/arvo/compare/maint-20160818...joemfb:door-prelude

```
|=  arg/@
=<  abet:(~(flub . 10) arg)
|/  a/@
+*  ~|  %too-big
    ?>  (lth a 20)
    =*(this . .)
++  abet  [~ a]
++  flub
  |=  b/@
  this(a (add a b))
--
```

```
> +test 9
[~ 19]
> +test 10
/~zod/home/0/gen/test:<[2 1].[11 3]>
/~zod/home/0/gen/test:<[2 5].[2 28]>
/~zod/home/0/gen/test:<[4 5].[6 17]>
/~zod/home/0/gen/test:<[5 5].[6 17]>
%too-big
```

### custom aliases

https://github.com/urbit/arvo/compare/maint-20160818...joemfb:door-alias

```
|=  arg/@
=<  abet:(~(flub . 10) arg)
|/  a/@
+*  this  .
    samp  +<
    cont  +>
++  abet  [~ a]
++  flub
  |=  b/@
  ?>  =(a samp)
  ?>  =(+>+>.$ cont)
  this(a (add a b))
--
```

```
> +test 10
[~ 20]
```

**Comment by joemfb** [comment]
The magic-face example is somewhat fragile:

```
[%vega-start [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/hoon]]
%vega-parsed
[%vega-compiled %151 151]
%hoon-load
[%tang [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/zuse] ~nocsyr-ribnym]
[%vane %a [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/ames] ~fidsen-pitfed]
find-fork-e
-find.abet
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,517 53].[1,517 61]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,517 45].[1,517 62]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,517 22].[1,517 63]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,515 11].[1,518 13]>
-find.zork
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,082 21].[1,082 61]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,082 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,079 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,078 11].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,078 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,077 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,076 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,075 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[1,074 7].[1,083 41]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[969 5].[1,551 7]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[967 3].[1,552 5]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[4 7].[1,553 8]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[4 3].[1,896 5]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/ames:<[3 3].[1,896 5]>
[%vane %b [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/behn] ~finweg-pandys]
[%vane %c [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/clay] ~dirtun-mipfyr]
[%vane %d [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/dill] ~tortex-lanmud]
find-fork-e
-find.abet
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[488 19].[488 66]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[488 5].[490 56]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[485 5].[490 56]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[483 5].[490 56]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[481 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[480 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[472 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[470 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[460 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[459 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[456 3].[496 13]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[454 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[102 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[101 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[100 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[41 5].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[37 5].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[8 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[5 1].[537 3]>
/~zod/home/~2017.3.7..02.29.30..9481/arvo/dill:<[4 1].[537 3]>
[%vane %e [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/eyre] ~mispun-dasnel]
[%vane %f [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/ford] ~sopnem-mosnel]
[%vane %g [i=~.~zod t=/home/~2017.3.7..02.29.30..9481/arvo/gall] ~ligwep-hinfer]
: /~zod/home/31/arvo/hoon/hoon
<<<reset>>>
```

Maybe related: `$aka` with `$like` is, uh, not right ...

```
> =<(foo =+(a=3 =*(foo a .)))
3
> =<(foo =+(a=3 ^+(. =*(foo a .))))
< 1.dsa
  { c/{p/{p/@ud q/@ud} q/""}
    b/<1.dut {tub/{p/{p/@ud q/@ud} q/""} <1.bge {daf/@tD <409.yxa 110.lxv 1.ztu $151>}>}>
      a
    < 1.dsa
      { c/{p/{p/@ud q/@ud} q/""}
        b/<1.dut {tub/{p/{p/@ud q/@ud} q/""} <1.bge {daf/@tD <409.yxa 110.lxv 1.ztu $151>}>}>
        a/<1.dut {tub/{p/{p/@ud q/@ud} q/""} <1.bge {daf/@tD <409.yxa 110.lxv 1.ztu $151>}>}>
        v/<288.wmn 42.uch 409.yxa 110.lxv 1.ztu $151>
      }
    >
    v/<288.wmn 42.uch 409.yxa 110.lxv 1.ztu $151>
  }
```

**Comment by cgyarvin** [comment]
Yeah the nice thing is that none of these things is hard to implement -- so
long as you do them at parse time.

The magic face is just too hokey for me.  I'm sorry, magic face.

I like the custom aliases.  It's cute that of course you can do these
without a terminator.  The custom aliases are also less powerful than a
general-purpose prelude, an obviously desirable quality -- unless someone
can provide another use case for a general-purpose prelude...

I do enjoy the grim binary discipline of +>.$, etc.  But maybe Hoon already
has enough of this.  We may already be working hard enough at being the Joy
Division of programming languages.

**Comment by joemfb** [comment]
The custom alias strikes me as the most acceptable/least disruptive, although the magic face has grown on me.

With regards to the custom subject, @chc4 provided an app access control use-case above, but I'm not sure that you'd really want that on *every* arm... The best example I can think of is validation in a data-structure engine -- it's easy to image `+*  (apt a)` in `++in` and `++by`. Or something similar in userspace engines ... That's cool, but is it cool enough to demand syntactic accommodation?

The only other option I can think of is the partial-twig example that I posted earlier. (I didn't implement that for obvious reasons.). That would be really cool, in that it almost rises to the level of a user macro. On the other hand, it almost rises to the level of a user macro ...

**Comment by cgyarvin** [comment]
You definitely wouldn't want to validate the whole treap on every use of
++in and ++by!

This is in a way the problem with my design (full prelude) -- it's
overpowered and doesn't limit itself to the set of thing you'd actually
want to do.  I say, PR the custom alias...

**Comment by chc4** [comment]
Wow, that's awesome that you sketched out all of them. I mentioned it before, but because cores are concatative you can use =< to have preludes on all the arms of an interior core, and none on the one you are actually returning. I will admit that adding a mechanism for this in Hoon is probably too heavy, though. (Typing this out makes me think that it's actually perfectly possible to do the access control stuff with vanilla Hoon...)

Is the plan then to change %door so that it expands into a %dork with all those aliases by default, or just expect people to copy-paste that when constructing a state machine?

And, uh, consider this an objection on naming it dork please ;_; crux maybe, for the core allusion?

**Comment by joemfb** [comment]
`$dork` is just temporary ;)  There's two `:door` runes: `|_` and `|/`. I just appropriated the second and gave it a new twig book for easy testing/resetting. My PR will just be a modification to `:door` to support the optional aliases.

Speaking of which, anybody know how to `|reset` with a new `++twig` that doesn't nest in the old?

**Comment by cgyarvin** [comment]
At present I'm not sure this reset works even in theory!  It's supposed to work in the new arvo, though.

If it does work in old arvo it involves decrementing the Hoon version number.  This is not just cosmetic... but for development I wouldn't mess with it.  Just make a pill.

**Comment by joemfb** [comment]
So, about that... I can't seem to make a pill either. Here's what I've done:

- disable |autoload
- copy in hoon.hoon (with `$dork` moved to `$door`)
- +solid or +metal

In either case, I get a nest-fail at the actual compilation step:

```
> .door/pill +metal ~zod
...
/~zod/home/0/gen/metal:<[167 24].[167 62]>
nest-fail
```

```
> .door/pill +solid
...
/~zod/home/0/gen/solid:<[21 11].[21 39]>
nest-fail
```

What am I missing?

**Comment by joemfb** [comment]
While am at it, any thoughts on this?

```
> =/(a *?($ash $elm) ^-(foot [a *twig]))
nest-fail
> =/(a *?($ash $elm) ^-(foot ?:(?=($ash a) [a *twig] [a *twig]))
[%ash [%$ p=0]]
```

**Comment by cgyarvin** [comment]
Look at gen/metal.hoon -- it's certainly the best-commented file in Urbit.

Your compiler isn't compiling, I don't know why!  Put a !: in hoon.hoon...

**Comment by cgyarvin** [comment]
That's pretty curious.  One would have to dig into it...

**Comment by ohAitch** [comment]
This is definitely a thing I've run into before, basically the compiler
isn't smart enough to fuse `[?(a b) foo]` with `?([a foo] [b foo])`

**Comment by ohAitch** [comment]
```
> `$%({$& $~} {$| $~})`*{? $~}
/~zod/home/~2017.3.2..20.42.50..0b14/sys/vane/ford:<[1.207 24].[1.207 52]>
-need.?({$.n $~} {$.y $~})
-have.{?($.n $.y) $~}
nest-fail
```

**Comment by cgyarvin** [comment]
Yeah, that's correct.  It was never advertised as the world's smartest type
system.

**Comment by joemfb** [comment]
My bad, I assumed the compilation error was due to `++twig` itself nest-failing, but there are actual errors ...

**Comment by hoclun-rigsep** [comment]
This discussion may have gotten deeper than where I can as yet swim, but, assuming my novice perspective is of any value: this issue seems an overreaction to our aversion to typing the same word more than once—`++  arm-name`  with `..arm-name` to follow (usually very closely to follow, in a cast at the top of the arm...). I admit that at first I looked for ways to not-repeat myself but before you knew it typing `..arm-name` became no big deal. If you end up changing the name of the arm and not the subsequent reference, the resulting error is a crystal clear `find`.

**Comment by juped** [comment]
hoclun is absolutely 301% correct in this

**Comment by joemfb** [comment]
I think the big advantage comes with defining aliases in a door.

This has become a fairly common pattern in some of the new vanes:

```
=|  {some/^ faces/@}
=*  the-whole-sample +<
|%
++  my-arm  ...
--
```

And it works fine, but it doesn't *look* like a door. At least not until you're comfortable with how the various core twigs are expanded. Compare:

```
|_  {some/^ faces/@}
+*  the-whole-sample +<
++  my-arm  ...
--
```

Being able to add a `this` alias for the whole core is a nice additional benefit.

**Comment by joemfb** [comment]
Of course, most of the code just uses `+<` directly. And the addresses get pretty opaque once you nest cores ...

**Comment by chc4** [comment]
Yup, the reasoning behind this is more for pattern matching to recognize "hey, this is a state machine" than otherwise. Like I said before, I don't really have as much aversion to `..arm` other than that it's not clear what it actually means, and that it has no relation to `arm` itself.

And yeah, /people don't actually use it/ which is annoying.

**Comment by hoclun-rigsep** [comment]
@joemb, why don't people just put a face on the whole sample: `sample/{some/^ faces/@}` ? I'm sure I've seen that all over the code too.

**Comment by joemfb** [comment]
I tend to do that, but sometimes limbs get pretty long. And there's a lot of code that doesn't

 I think the best use-case for this is an app with a large, complex state mold. You only ever refer to the whole sample when casting or producing the product of your arms, the rest of the time you only want a subtree. Not having to indirect through a face for the whole sample can make things cleaner.

**Comment by ohAitch** [comment]
I think in this case you want something like  sample,/{some/^ faces/@},
where ,/ means "both put and don't put this face on", i.e. references to
"sample" resolve correctly but a plain "some" reference into it also works.

**Comment by belisarius222** [comment]
@joemfb One of three commits in https://github.com/urbit/arvo/pull/365 got merged. How does that affect the status of this issue? I confess I don't follow this discussion fully.

**Comment by joemfb** [comment]
Looks to me like all three commits got merged (per `git branch --contains <hash>`). So that means `+*` pseudo-arms are available and this feature-request/use-case is covered by `+*  this  .` ...

**Comment by Fang-** [comment]
Yeah, those seem to work. Cool, closing then. (:
